src/ensembl/doc/architecture/decisions/0016-server-side-rendering.md

```diff
@@ -0,0 +1,149 @@
+
+##### Example
+As of September 2020, HGNC's site `genenames.org` is an Angular-based site, with its metadata tags filled in dynamically in the browser. The html layout returned from the server has the following metatags (note that they are empty and contain Angular bindings for the data that is loaded on the client side):
+
+```html
+<title ng-bind="vm.headTitle"></title>
+<meta name="Description" content="{[{vm.metaDescription}]}" />
+```
+
+A sample query `brca2 site:https://www.genenames.org/` in different search engines yields the following results:
+- Google shows the relevant page (`https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:1101`) in the list of search results, with correct values read from the title and description tags after they have been populated on the client side.
+- DuckDuckGo (uses Bing search engine) cannot find the page.
+
+This experiment confirms that Google is capable of indexing client-side-rendered web pages, but that other search engines struggle. With server-side-rendered pages, all search engines should be perfectly capable of indexing them.
+
+#### Lack of contextual information for 3rd-party bots
+When someone shares a url to a web page on social media or in a messaging app such as Slack, this service will often send a web crawler to the shared page to scrape its data (such as its title, description, and possibly an image) for a preview. These aren't sophisticated crawlers, and they do not execute javascript; so with client-side rendering, they will not be able to retrieve the title and the description of the shared page.
+
+### Additional advantages of server-side rendering
+The points below are expected to offer only a marginal benefit in the context of the new Ensembl website, and are mentioned only for the sake of completeness.
+
+#### Higher page reliability
+With client-side rendering, if anything happens to the javascript bundle, the page will not render. In contrast, with server-side rendering, the browser will display the contents of the page in the way it was rendered on the server.
+
+_**However**, we recognize that in the case of the Ensembl site, there is little value in being able to view the page, but not interact with it._
+
+#### Perception of faster page load (not guaranteed)
+While the time to the first byte with server-side rendering will inevitably be slower than with client-sidee rendering, it is, in principle, possible that the time it takes to build a page on the server and deliver it to the browser is less than what it would take a client-rendered page to fetch the script, then fetch the data, and then finally display the page content.
+
+_**However**, we do recognize that in practice, these gains, if at all attainable, will be negligible. We expect that Ensembl users are bioinformaticians or clinicians accessing Ensembl from office computers, which are sufficiently powerful and are on sufficiently fast networks that performance gains will not be meaningful._
```

Comment:
LMIC users might not be in as such a good place with their machines or internet

Reply:
Absolutely, but the way we are building ensembl-2020 is effectively disregarding LMICs.

- we are not, while building the site, routinely checking its performance on lower-end hardware
- we are not checking its performance on slow connections
- we know that for some applications, e.g. genome browser, we are expecting the client to have a pretty fast connection (lots of data to download)

If LMIC are really a concern, something needs to change about our development process. But I believe our target audience, even in such countries, are on reasonably powerful desktops/laptops and on a decent internet connection rather than on underpowered phones on patchy 3G connection. For the latter group, which Google likes to describe as a likely target for modern web developers, our site will perform poorly.

Reply:
I'm not sure it is disregarding LMICs per-say, but it is discounting those who have the issues you outlined and they may disproportionally affect LMIC researchers. Point 1 is a hard one to resolve so I'm not going to address this.

Points 2 and 3 though. Dodgy network connections can happen anywhere. I have them time to time in my own home. It is a fact of life not the exception. Data locality will help and drives the point for CDN like behaviour. Also whilst yes we will be shipping a lot of data to the client, I've not seen the volumes to be excessive to the point I'm worried.

Reply:
> Dodgy network connections can happen anywhere. I have them time to time in my own home. It is a fact of life not the exception.

Yes, I agree.

Server-side rendering brings with it a couple of UI-related concerns that I can think of:
- Time between when the contents of the page appear in the browser (defined by when the HTML document reaches the browser) and when they become interactive (extra time for a round trip for javascript). On slow networks, this may become noticeable and is an unpleasant experience [see the demo](https://youtu.be/0A-2BhEZiM4?t=337)
- Some of our data that's important for rendering is stored on the client side and is unavailable on the server (e.g. the list of selected species that we show as species lozenges in the top area of the site).

It is likely then that we will only be able to introduce some minimal aspects of server-side rendering, such as populating the `<head>` of the html document with relevant data and maybe showing a rough skeleton of the page, which we will then fill with data when javascript is loaded and relevant data has been requested. It should be better than looking at a white screen anyway. We need to bring Andrea on board with this additional design complexity.

cc @ens-st3